File: 2021/10/31/main.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        if head == None:
            return None
        curr = head
        while curr != None:
            if curr.child == None:
                curr = curr.next
                continue

            child = curr.child
            while child.next != None:
                child = child.next
            child.next = curr.next
            if curr.next != None:
                curr.next.prev = child
            
            curr.next = curr.child
            curr.child.prev = curr
            curr.child = None
        
        return head

File: 2021/10/31/main.go
package main

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Prev *Node
 *     Next *Node
 *     Child *Node
 * }
 */

func flatten(root *Node) *Node {
	if root == nil {
		return nil
	}
	curr := root
	for curr != nil {
		if curr.Child == nil {
			curr = curr.Next
			continue
		}

		child := curr.Child
		for child.Next != nil {
			child = child.Next
		}
		child.Next = curr.Next
		if curr.Next != nil {
			curr.Next.Prev = child
		}
		curr.Next = curr.Child
		curr.Child.Prev = curr
		curr.Child = nil
	}
	return root
}
